### ISSUE TITLE
[tt-train] Optimizer state isn't per parameter: wrong AdamW update size for a late-starting parameter, and a parameter frozen at construction crashes or never trains

### ISSUE THREAD

**Issue by Patriot117** [tenstorrent/tt-metal#57752]
### Component / Area

tt-train (optimizers)

### Issue Type (optional)

Bad Outputs

### Observed

There are two related problems, and both come from tracking optimizer state for the whole optimizer instead of per parameter. Part 1 is in the four AdamW optimizers (AdamW, AdamWFullPrecision, MorehAdamW, AdamWComposite) and is the AdamW counterpart of the SGD late-momentum issue fixed in #53222. Part 2 is in those four plus MuonComposite. SGDComposite has the same part 2 gap, but #54601 already plans that fix, so it's left out here.

**1. Bias correction uses a global step count.** `m_steps` and `m_beta1_pow` / `m_beta2_pow` advance once per `step()` for the whole optimizer. That happens at `optimizers/adamw.cpp:74-77`, `adamw_full_precision.cpp:62-65`, `adamw_composite.cpp:63` for MorehAdamW, and `:181` for AdamWComposite (its bias correction at `:210-230` uses `std::pow(beta, m_steps)`). A parameter that was skipped on some steps because it had no gradient still gets bias correction for the global count. torch.optim.AdamW instead keeps `state["step"]` per parameter and only advances it when that parameter has a gradient. So a parameter whose first gradient arrives late gets updates of the wrong size. If its first gradient arrives on the third step, its first update is about 0.64x what torch gives (with default betas, `(1-b1)/(1-b1^3)` divided by `sqrt((1-b2)/(1-b2^3))`). After 1,000 or more skipped steps, its early updates are 2.5x to 6.5x torch's, and they are still 1.17x to 1.26x after 1,000 of its own updates.

**2. State is created only for parameters that require grad at construction.** Each constructor emplaces moment buffers only when `get_requires_grad()` is true (`adamw.cpp:43-55`, `adamw_full_precision.cpp:25-42`, `adamw_composite.cpp:33-45` and `:136-149`). If a parameter is frozen when the optimizer is built and later unfrozen:
- AdamW and AdamWFullPrecision iterate `m_parameters` in `step()` and call `m_exp_avg.at(name)` (`adamw.cpp:87-88`, `adamw_full_precision.cpp:74-76`), which throws `std::out_of_range`.
- MorehAdamW and AdamWComposite iterate their moment maps instead, so the parameter is never visited. It never trains, and no error is raised.
- MuonComposite fails the same silent way. Its constructor creates `m_momentum_buffer` only for parameters that require grad (`muon_composite.cpp:34-52`), and `step()` iterates that map (`muon_composite.cpp:68`).

Fused SGD already gets this right: its `step()` iterates `m_parameters` (`sgd.cpp:49-52`) and creates a missing momentum buffer on first use (`sgd.cpp:58-64`). #54601 proposes the same lazy creation for SGDComposite, whose `step()` iterates `m_theta` and so can't see a parameter without a buffer. It doesn't cover the AdamW optimizers or MuonComposite, which have the same shape.

torch.optim AdamW, SGD and Muon all create state on a parameter's first gradient (`adam.py` `_init_group`, `sgd.py:88-102`, `_muon.py:146-164`), so every one of these cases works there.

### Expected

Match torch.optim. Each AdamW parameter's bias correction uses the number of steps in which that parameter had a gradient. Every optimizer creates state on a parameter's first gradient, as fused SGD already does, so a parameter unfrozen after construction trains normally.

### Suggested fix

Mirror what #53222 did for SGD:
- Keep the step count, or `beta1_pow` / `beta2_pow`, per parameter (for example a map keyed by name next to `m_exp_avg`), advance it only when that parameter is updated, and pass the per-parameter values to the kernel.
- Create the moment buffers lazily in `step()` the first time a parameter has an initialized gradient, instead of only in the constructor. `SGD::step` already does this (`sgd.cpp:58-64`). For MuonComposite that means iterating `m_parameters` in `step()` and creating `m_momentum_buffer` entries on first use, the change #54601 proposes for SGDComposite.
- Both changes touch `get_state_dict` / `set_state_dict`, so old checkpoints need a fallback. Seeding every per-parameter count from the saved global `steps` matches today's behavior for parameters that trained every step.
- Lazy creation runs into the checkpoint blocker described in #54601: `read_named_parameters` iterates the live map, so an optimizer that hasn't created an entry yet won't restore it from the file. Whatever reader change lands for SGD there would cover these optimizers too.

### 1. Steps (exact commands)

1. Download the two attached scripts, `tt_adamw_step_repro.py` and `tt_adamw_lazy_state_repro.py`.
2. Run `python3 tt_adamw_step_repro.py` and `python3 tt_adamw_lazy_state_repro.py` (need only numpy).

The two attached Python scripts copy the control flow and don't call ttml.

`tt_adamw_step_repro.py`: parameters A and B, B has no gradient for K steps, then both train for 1,000 steps. It compares B's own update size under per-parameter (torch) vs global (ttml) bias correction.

`tt_adamw_lazy_state_repro.py`: B frozen when the optimizer is built, unfrozen before the next step.

gtest sketch following `run_late_momentum_case` in `tt-train/tests/optimizers/sgd_test.cpp` (not compiled; the fixture is the same shape as `SGDLateMomentumTest`):

```cpp
// "late" gets its first gradient on the third step. With per-parameter bias correction its
// first update is lr * g / (|g| + eps), i.e. about lr per element for g in [0.25, 1].
// With the current global step count it is about 0.64 * lr.
template <typename Optimizer>
static void run_adamw_late_param_case() {
    using namespace ttml;

    const std::array<std::size_t, 4> shape = {1, 1, 32, 32};
    const float lr = 0.1f;

    autograd::ctx().set_seed(123U);
    auto& gen = autograd::ctx().get_generator();
    xt::xarray<float> w_early = test_utils::make_uniform_xarray<float>(shape, -1.0F, 1.0F, gen());
    xt::xarray<float> g_early = test_utils::make_uniform_xarray<float>(shape, -1.0F, 1.0F, gen());
    xt::xarray<float> w_late = xt::zeros<float>(shape);
    xt::xarray<float> g_late = test_utils::make_uniform_xarray<float>(shape, 0.25F, 1.0F, gen());

    auto theta_early = autograd::create_tensor(to_tt(w_early), true);
    auto theta_late = autograd::create_tensor(to_tt(w_late), true);
    ttml::serialization::NamedParameters params{{"early", theta_early}, {"late", theta_late}};

    optimizers::AdamWConfig cfg;
    cfg.lr = lr;
    cfg.weight_decay = 0.0f;
    Optimizer opt(params, cfg);

    theta_early->set_grad(to_tt(g_early));
    opt.step();
    opt.step();

    theta_late->set_grad(to_tt(g_late));
    opt.step();

    auto late = ttml::core::to_xtensor(theta_late->get_value());
    for (size_t i = 0; i < late.size(); ++i) {
        EXPECT_NEAR(late.flat(i), -lr, 0.05f * lr) << "first update must use the parameter's own step count";
    }
}

// Frozen at construction, unfrozen before the first step: must train, not throw or skip.
template <typename Optimizer, typename Config>
static void run_unfreeze_after_construction_case() {
    using namespace ttml;

    const std::array<std::size_t, 4> shape = {1, 1, 32, 32};
    autograd::ctx().set_seed(123U);
    auto& gen = autograd::ctx().get_generator();
    xt::xarray<float> w = xt::zeros<float>(shape);
    xt::xarray<float> g = test_utils::make_uniform_xarray<float>(shape, 0.25F, 1.0F, gen());

    auto theta = autograd::create_tensor(to_tt(w), false);
    ttml::serialization::NamedParameters params{{"theta", theta}};
    Config cfg;
    cfg.lr = 0.1f;
    Optimizer opt(params, cfg);

    theta->set_requires_grad(true);
    theta->set_grad(to_tt(g));
    EXPECT_NO_THROW(opt.step());
    auto after = ttml::core::to_xtensor(theta->get_value());
    EXPECT_FALSE(after == w) << "an unfrozen parameter must be updated";
}
```

Run the first with `AdamW`, `AdamWFullPrecision`, `MorehAdamW` and `AdamWComposite`. Run the second with those four plus `MuonComposite` / `MuonConfig`. Going by the source, fused `SGD` / `SGDConfig` should already pass the second one.

### 2. Input data / link or description

None needed. The scripts build their own random tensors from a fixed seed. No dataset, model or config.

### 3. Frequency

Every run. The scripts use a fixed seed and their output is identical each time.

### 1. Software Versions

OS version: macOS 27.0 (the scripts are plain Python and don't depend on the OS)
Software / firmware release version: tt-metal main at c87cf141a5 (source read, not built). Scripts run with Python 3.14.7 and numpy 2.5.2.

### 2. Hardware Details

Product: none. Not run on Tenstorrent hardware.
Card/System: none. Found by reading source and reproduced with the two attached standalone numpy scripts.

### Is this a regression?

Unknown

### Logs & Diagnostics

Output of `tt_adamw_step_repro.py`:

```
Sanity (force-green): no freezing, the two emulations must agree exactly

K = 0 frozen steps, then 1000 steps with gradients for B
  A (never frozen) max |torch - ttml| param diff: 0.000e+00
  B own update #   1: expected (torch) |step| = 1.000e-03   actual (ttml) = 1.000e-03   ratio = 1.00x
  B own update #  10: expected (torch) |step| = 6.547e-04   actual (ttml) = 6.547e-04   ratio = 1.00x
  B own update # 100: expected (torch) |step| = 1.598e-04   actual (ttml) = 1.598e-04   ratio = 1.00x
  B own update #1000: expected (torch) |step| = 7.967e-04   actual (ttml) = 7.967e-04   ratio = 1.00x
  B final max |torch - ttml| param diff: 0.000e+00  (torch |B| max 3.872e-02)

Bug case: B unfrozen late

K = 10 frozen steps, then 1000 steps with gradients for B
  A (never frozen) max |torch - ttml| param diff: 0.000e+00
  B own update #   1: expected (torch) |step| = 1.000e-03   actual (ttml) = 4.821e-04   ratio = 0.48x
  B own update #  10: expected (torch) |step| = 5.497e-04   actual (ttml) = 5.750e-04   ratio = 1.05x
  B own update # 100: expected (torch) |step| = 2.524e-04   actual (ttml) = 2.641e-04   ratio = 1.05x
  B own update #1000: expected (torch) |step| = 3.396e-04   actual (ttml) = 3.406e-04   ratio = 1.00x
  B final max |torch - ttml| param diff: 2.244e-03  (torch |B| max 3.343e-02)

K = 1000 frozen steps, then 1000 steps with gradients for B
  A (never frozen) max |torch - ttml| param diff: 0.000e+00
  B own update #   1: expected (torch) |step| = 1.000e-03   actual (ttml) = 2.515e-03   ratio = 2.52x
  B own update #  10: expected (torch) |step| = 5.617e-04   actual (ttml) = 2.924e-03   ratio = 5.21x
  B own update # 100: expected (torch) |step| = 4.666e-04   actual (ttml) = 1.235e-03   ratio = 2.65x
  B own update #1000: expected (torch) |step| = 4.620e-04   actual (ttml) = 5.403e-04   ratio = 1.17x
  B final max |torch - ttml| param diff: 8.255e-02  (torch |B| max 7.164e-02)

K = 10000 frozen steps, then 1000 steps with gradients for B
  A (never frozen) max |torch - ttml| param diff: 0.000e+00
  B own update #   1: expected (torch) |step| = 1.000e-03   actual (ttml) = 3.162e-03   ratio = 3.16x
  B own update #  10: expected (torch) |step| = 4.962e-04   actual (ttml) = 3.239e-03   ratio = 6.53x
  B own update # 100: expected (torch) |step| = 2.978e-04   actual (ttml) = 9.650e-04   ratio = 3.24x
  B own update #1000: expected (torch) |step| = 2.510e-04   actual (ttml) = 3.157e-04   ratio = 1.26x
  B final max |torch - ttml| param diff: 7.781e-02  (torch |B| max 6.680e-02)

RESULT: FAIL - 3/3 late-unfreeze cases give a first update for B that differs from torch.optim.AdamW by more than 5%
```

Output of `tt_adamw_lazy_state_repro.py`:

```
Sanity: nothing frozen at construction, every class updates both parameters
  OK

Bug case: B frozen at construction, unfrozen before the next step
  expected (torch.optim AdamW, SGD, Muon) params updated: ['A', 'B']
  actual AdamW / AdamWFullPrecision: throws unordered_map::at: key 'B' not found
  actual MorehAdamW / AdamWComposite: updated ['A'] (B has a gradient but is skipped, no error)
  actual MuonComposite: updated ['A'] (B has a gradient but is skipped, no error)
  control, ttml SGD (fused): updated ['A', 'B']

RESULT: FAIL - 3/3 optimizer shapes diverge from torch.optim; ttml's own fused SGD does not
```

[tt_adamw_step_repro.py](https://github.com/user-attachments/files/32620024/tt_adamw_step_repro.py)
[tt_adamw_lazy_state_repro.py](https://github.com/user-attachments/files/32620025/tt_adamw_lazy_state_repro.py)

### Priority

None

### Impact

Affected workflows: AdamW training where a parameter gets its first gradient late (its updates are the wrong size, with no error), and AdamW or Muon training where a parameter is frozen when the optimizer is built and unfrozen later (it never trains, or step() throws std::out_of_range).
